src/g_editor.c

```diff
@@ -4135,38 +4139,58 @@ static void canvas_paste(t_canvas *x)
     }
 }
 
-static void canvas_paste_from_clipboard_text(t_canvas *x)
-{
-    pdgui_vmess("pdtk_get_clipboard_text", "^", x);
-}
+static t_binbuf *clipboard_patch_bb = NULL;
 
-void unescape_clipboard_patch(char *str) {
-    char *temp = malloc(strlen(str) + 1);
-    if (!temp) return;
-    int i = 0, j = 0;
-    while (str[i]) {
-        if (str[i] == ';' && str[i+1] == 'n') {
-            temp[j++] = ';';
-            temp[j++] = '\n';
-            i += 2;
-        } else {
-            temp[j++] = str[i++];
-        }
+void canvas_got_clipboard_contents(t_canvas *x, t_floatarg flagf, t_symbol *s) {
+    int flag = flagf;
+    
+    switch (flag) {
+        case CLIPBOARD_PATCH_TEXT_START:
+            if (clipboard_patch_bb) {
+                binbuf_free(clipboard_patch_bb);
+            }
+            clipboard_patch_bb = binbuf_new();
+            break;
+
+        case CLIPBOARD_PATCH_TEXT_LINE_END:
+            if (clipboard_patch_bb) {
+                char* received_data = strdup(s->s_name);
+                t_binbuf *temp_bb = binbuf_new();
+                binbuf_text(temp_bb, received_data, strlen(received_data));
+                binbuf_addsemi(temp_bb);
+                binbuf_add(clipboard_patch_bb, binbuf_getnatom(temp_bb), binbuf_getvec(temp_bb));
+                binbuf_clear(temp_bb);
+                binbuf_free(temp_bb);
+                free(received_data);
+            }
+            break;
+
+        case CLIPBOARD_PATCH_TEXT_LINE_PARTIAL:
+            if (clipboard_patch_bb) {
+                char* received_data = strdup(s->s_name);
+                t_binbuf *temp_bb = binbuf_new();
+                binbuf_text(temp_bb, received_data, strlen(received_data));
+                binbuf_add(clipboard_patch_bb, binbuf_getnatom(temp_bb), binbuf_getvec(temp_bb));
+                binbuf_clear(temp_bb);
+                binbuf_free(temp_bb);
+                free(received_data);
+            }
+            break;
+
```

Comment:
is the `CLIPBOARD_PATCH_TEXT_LINE_PARTIAL` for future use? right now it looks suspiciously identical to `CLIPBOARD_PATCH_TEXT_LINE_END`.

also, i'm not sure why you need to `strdup` the incoming data

Reply:
ah, i missed the `binbuf_addsemi()`...

Reply:
I'm doing case fall-through now with conditional addition of semi next